File: jee/java/src/main/java/fables/kotlin/jee/rest/KittenRestService.java
package fables.kotlin.jee.rest;

import fables.kotlin.jee.business.KittenBusinessService;
import fables.kotlin.jee.business.KittenEntity;

import javax.inject.Inject;
import javax.ws.rs.*;

/**
 * JSON REST CRud service.
 * JEE will first create one noarg instance, and then injected instances.
 *
 * @author Zeljko Trogrlic
 */
@Path("kitten")
public class KittenRestService {

  @Inject
  protected KittenBusinessService kittenBusinessService;

  @GET
  @Path("{id}")
  @Produces({"application/json"})
  public KittenRest find(
      @PathParam("id") final int id
  ) {
    return kittenBusinessService
        .find(id)
        .map(kittenEntity -> new KittenRest(kittenEntity.getName(), kittenEntity.getCuteness()))
        .orElseThrow(() -> new NotFoundException("ID " + id + " not found"));
  }

  @POST
  @Produces({"application/json"})
  public Integer add(KittenRest kittenRest) {
    KittenEntity kittenEntity = new KittenEntity(kittenRest.getName(), kittenRest.getCuteness());
    return kittenBusinessService.add(kittenEntity);
  }
}

File: jee/plugins/src/main/kotlin/fables/kotlin/jee/rest/KittenRestService.kt
package fables.kotlin.jee.rest

import fables.kotlin.jee.business.KittenBusinessService
import fables.kotlin.jee.business.KittenEntity
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType

/**
 * JSON REST CRud service.
 * Class and all public methods are opened by Kotlin compiler plugin.
 * JEE will first create one noarg constructor instance, and then injected constructor instances.
 * @author Zeljko Trogrlic
 */
@Path("kitten")
class KittenRestService @Inject constructor(private val kittenBusinessService: KittenBusinessService) {

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    fun find(
            @PathParam("id") id: Int
    ): KittenRest = kittenBusinessService
            .find(id)
            .map { kittenEntity -> KittenRest(kittenEntity.name, kittenEntity.cuteness) }
            .orElseThrow { NotFoundException("ID $id not found") }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    fun add(
            kittenRest: KittenRest
    ): Int {
        val kittenEntity = KittenEntity(kittenRest.name, kittenRest.cuteness)
        return kittenBusinessService.add(kittenEntity)
    }
}
